### PR TITLE
[BEAM-6361][BEAM-6364] fix user-metric-prefix checking in Flink portable metrics update

### PR DESCRIPTION
**Pull request by ryan-williams**
two small follow-ups to #7183:
- [BEAM-6361](https://issues.apache.org/jira/browse/BEAM-6361): fix incorrect user-metric-prefix check
- [BEAM-6364](https://issues.apache.org/jira/browse/BEAM-6364): WARN instead of `throw` on as-yet-unsupported metrics types

They're each pretty small, and the tests of the latter depend on the former, so I put them both in this PR; let me know if it would be better if I split them out.

R: @ajamato 
R: @robertwb (we'd previously discussed the [BEAM-6364](https://issues.apache.org/jira/browse/BEAM-6364) portion of this in particular)

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/flink/src/test/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainerTest.java [hunk 1]
```diff
@@ -106,6 +110,39 @@ public void testGauge() {
...
+  @Test
+  public void testMonitoringInfoUpdate() {
+    FlinkMetricContainer container = new FlinkMetricContainer(runtimeContext);
+    MetricsContainer step = container.getMetricsContainer("step");
+
+    SimpleCounter userCounter = new SimpleCounter();
+    when(metricGroup.counter("ns1.metric1")).thenReturn(userCounter);
+
+    SimpleCounter elemCounter = new SimpleCounter();
+    when(metricGroup.counter("beam.metric:element_count:v1")).thenReturn(elemCounter);
```
Comment (ryan-williams) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainerTest.java]: The current namespace parsing logic [above](https://github.com/apache/beam/pull/7408/files#diff-bd906385f3e3ff74094985af80da5a41R109) just takes the everything after the first `:` as the "name", so this name is `metric:element_count:v1`.

I don't see any reason that having additional `:`s in the "name" is a problem, but I saw [a note about it (only related to user metrics, though) in SimpleMonitoringInfoBuilder](https://github.com/apache/beam/blob/5716dba6d10f32dfa8ab807ffacc2e68f9267225/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleMonitoringInfoBuilder.java#L114-L115), so open to whether there's a better way to think about it here.

**Comment by ryan-williams** [comment]
python [failure]() seems unrelated, and the same as in [3385](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3385/console) (#7416), [3382](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3382/console) (#7300), [3370](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3370/console) (#7407), and perhaps others

```
======================================================================
FAIL: test_error_traceback_includes_user_code (apache_beam.runners.portability.portable_runner_test.PortableRunnerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/fn_api_runner_test.py", line 473, in test_error_traceback_includes_user_code
    self.assertIn('second', message)
AssertionError: 'second' not found in 'Traceback (most recent call last):\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/fn_api_runner_test.py", line 466, in test_error_traceback_includes_user_code\n    p | beam.Create([0]) | beam.Map(first)  # pylint: disable=expression-not-assigned\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/pipeline.py", line 425, in __exit__\n    self.run().wait_until_finish()\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/portable_runner.py", line 316, in wait_until_finish\n    self._job_id, self._state, self._last_error_message()))\nRuntimeError: Pipeline job-7dd476c6-fc08-41b1-8516-86e218de84b1 failed in state FAILED: start <DataInputOperation receivers=[ConsumerSet[Create/Read/Impulse.out0, coder=WindowedValueCoder[BytesCoder], len(consumers)=1]]>\n'
-------------------- >> begin captured logging << --------------------
```

**Comment by ryan-williams** [comment]
Run Python PreCommit

**Comment by boyuanzz** [comment]
> [python failure](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3387/console) seems unrelated, and the same as in [3385](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3385/console) (#7416), [3382](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3382/console) (#7300), [3370](https://builds.apache.org/job/beam_PreCommit_Python_Commit/3370/console) (#7407), and perhaps others
> 
> ```
> ======================================================================
> FAIL: test_error_traceback_includes_user_code (apache_beam.runners.portability.portable_runner_test.PortableRunnerTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/fn_api_runner_test.py", line 473, in test_error_traceback_includes_user_code
>     self.assertIn('second', message)
> AssertionError: 'second' not found in 'Traceback (most recent call last):\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/fn_api_runner_test.py", line 466, in test_error_traceback_includes_user_code\n    p | beam.Create([0]) | beam.Map(first)  # pylint: disable=expression-not-assigned\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/pipeline.py", line 425, in __exit__\n    self.run().wait_until_finish()\n  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_Commit@2/src/sdks/python/apache_beam/runners/portability/portable_runner.py", line 316, in wait_until_finish\n    self._job_id, self._state, self._last_error_message()))\nRuntimeError: Pipeline job-7dd476c6-fc08-41b1-8516-86e218de84b1 failed in state FAILED: start <DataInputOperation receivers=[ConsumerSet[Create/Read/Impulse.out0, coder=WindowedValueCoder[BytesCoder], len(consumers)=1]]>\n'
> -------------------- >> begin captured logging << --------------------
> ```

This error is tracking in https://issues.apache.org/jira/browse/BEAM-6280.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleMonitoringInfoBuilder.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ private boolean validate() {
       if (split.size() != 5) {
         LOG.warn(
             "Dropping MonitoringInfo for URN {}, UserMetric namespaces and "
-                + "name cannot contain ':' characters.",
+                + "names cannot contain ':' characters.",
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleMonitoringInfoBuilder.java]: It seems like you're agreeing that my fix is better? I'd changed it from "namespaces and name" to "namespaces and names".

I also agree about the `s/,/./` on the comma.

However it seems like this whole warning was dropped in the meantime, so this is all moot!

**Comment by angoenka** [comment]
Run Portable_Python PreCommit

**Comment by mxm** [comment]
If it's only that test failing, it's fine to merge. All other tests still run.

**Comment by ryan-williams** [comment]
Looks like everything passed the second time around, so this should be good to go

**Pull request merged**